packages/@aws-cdk/aws-ec2/lib/cfn-init-elements.ts

```diff
@@ -0,0 +1,405 @@
+
+    if (options.platform !== InitPlatform.WINDOWS && this.options.waitAfterCompletion !== undefined) {
+      throw new Error(`Command '${this.command}': 'waitAfterCompletion' is only valid for Windows systems.`);
+    }
+
+    for (const handle of this.options.serviceRestartHandles ?? []) {
+      handle.addCommand(commandKey);
+    }
+
+    return {
+      config: {
+        [commandKey]: {
+          command: this.command,
+          env: this.options.env,
+          cwd: this.options.cwd,
+          test: this.options.test,
+          ignoreErrors: this.options.ignoreErrors,
+          waitAfterCompletion: this.options.waitAfterCompletion?.toSeconds(),
+        },
+      },
+    };
+  }
+
+}
+
+/**
+ * Options for an InitService
+ */
+export interface InitServiceOptions {
+  /**
+   * Enable or disable this service
+   *
+   * Set to true to ensure that the service will be started automatically upon boot.
+   *
+   * Set to false to ensure that the service will not be started automatically upon boot.
+   *
+   * @default - true if used in `InitService.enable()`, no change to service
+   * state if used in `InitService.fromOptions()`.
+   */
+  readonly enabled?: boolean;
+
+  /**
+   * Make sure this service is running or not running after cfn-init finishes.
+   *
+   * Set to true to ensure that the service is running after cfn-init finishes.
+   *
+   * Set to false to ensure that the service is not running after cfn-init finishes.
+   *
+   * @default - same value as `enabled`.
+   */
+  readonly ensureRunning?: boolean;
+
+  /**
+   * Restart service when the actions registered into the restartHandle have been performed
+   *
+   * Register actions into the restartHandle by passing it to `InitFile`, `InitCommand`,
+   * `InitPackage` and `InitSource` objects.
+   *
+   * @default - No files trigger restart
+   */
+  readonly serviceRestartHandle?: InitServiceRestartHandle;
+}
+
+/**
+ * A services that be enabled, disabled or restarted when the instance is launched.
+ */
+export class InitService extends InitElement {
+  /**
+   * Enable and start the given service, optionally restarting it
+   */
+  public static enable(serviceName: string, options: InitServiceOptions = {}): InitService {
+    const { enabled, ensureRunning, ...otherOptions } = options;
+    return new InitService(serviceName, {
+      enabled: enabled ?? true,
+      ensureRunning: ensureRunning ?? enabled ?? true,
+      ...otherOptions,
+    });
+  }
+
+  /**
+   * Disable and stop the given service
+   */
+  public static disable(serviceName: string): InitService {
+    return new InitService(serviceName, { enabled: false, ensureRunning: false });
+  }
+
+  /**
+   * Create a service restart definition from the given options, not imposing any defaults.
+   *
+   * @param serviceName the name of the service to restart
+   * @param options service options
+   */
+  public static fromOptions(serviceName: string, options: InitServiceOptions = {}): InitService {
```

Comment:
Why is this not named `restart()` in line with the other ones?

What does it mean to not impose any defaults?

Reply:
This is just the final catch-all to let the user do whatever they want. You can definitely argue this is redundant with the enable()/disable(). If you call fromOptions and don't explicitly set one of enabled or ensureRunning, it's effectively a no-op. Should we just remove it entirely and keep .enable and .disable? (Rico and I were never quite happy with this).

Reply:
I would say so. We can add something here when we know what exactly we need.